cni-plugin/linkerd-cni-validator/src/main.rs

```diff
@@ -0,0 +1,258 @@
+        socket
+    };
+
+    tokio::select! {
+        is_readable = stream.readable() => {
+            is_readable.context("cannot read off client socket")?;
+        }
+
+        () = tokio::time::sleep(timeout) => {
+            anyhow::bail!("timed-out ({:?}) waiting for socket to become readable", timeout);
+        }
+    };
+
+    let mut buf = [0u8; 8];
+    let read_sz = stream.read_exact(&mut buf).await?;
+    let resp = String::from_utf8(buf[..expected_resp.len()].to_vec())?;
+    debug!(redirect_response = %resp, bytes_read = %read_sz);
+    anyhow::ensure!(
+        resp == expected_resp,
+        "expected client to receive {:?}, got {:?} instead",
+        expected_resp,
+        resp
+    );
+    Ok(())
+}
+
+#[tracing::instrument(name = "outbound_server", skip(ready, shutdown))]
+async fn outbound_serve(
+    listen_addr: SocketAddr,
+    ready: Arc<Notify>,
+    shutdown: kubert::shutdown::Watch,
+    resp: String,
+) -> Result<()> {
+    let listener = TcpListener::bind(listen_addr)
+        .await
+        .expect("Failed to bind server");
+    info!("Listening for incoming connections");
+
+    ready.notify_one();
+
+    tokio::select! {
+        _ = accept(listener, resp).in_current_span() => unreachable!("`accept` function never returns"),
+        _ = shutdown.signaled() => debug!("Received shutdown signal")
+    }
+
+    Ok(())
+}
+
+async fn accept(listener: TcpListener, resp: String) {
+    loop {
+        let (mut stream, client_addr) = listener
+            .accept()
+            .await
+            .expect("Failed to establish connection");
+        info!("Accepted connection");
+        let rng_resp = resp.clone();
+        let _ = tokio::spawn(
+            async move {
+                let resp_bytes = rng_resp.as_bytes();
+                // We expect this write to complete instantaneously,
+                // a timeout is not needed here.
+                match stream.write_all(resp_bytes).await {
+                    Ok(()) => debug!(written_bytes = resp_bytes.len()),
+                    Err(error) => error!(%error, "Failed to write bytes to client"),
+                }
+            }
+            .instrument(tracing::info_span!("conn", %client_addr)),
+        );
+    }
+}
+
+pub fn parse_timeout(s: &str) -> Result<time::Duration> {
+    let s = s.trim();
+    let (magnitude, unit) = if let Some(offset) = s.rfind(|c: char| c.is_digit(10)) {
+        let (magnitude, unit) = s.split_at(offset + 1);
+        let magnitude = magnitude.parse::<u64>()?;
+        (magnitude, unit)
+    } else {
+        anyhow::bail!("{} does not contain a timeout duration value", s);
+    };
+
+    let mul = match unit {
+        "" if magnitude == 0 => 0,
+        "ms" => 1,
+        "s" => 1000,
+        "m" => 1000 * 60,
+        "h" => 1000 * 60 * 60,
+        "d" => 1000 * 60 * 60 * 24,
+        _ => anyhow::bail!("invalid duration unit {}", unit),
```

Comment:
Great suggestion! I've also changed the error message below (when converting to ms) to better enunciate that the operation would overflow (instead of saying it's an invalid value)